src/api.rs

```diff
@@ -126,29 +188,147 @@ impl Account {
+            Account::Display(v) => v.to_owned(),
+            Account::Email(v) => v.to_owned(),
+            Account::Legal(v) => v.to_owned(),
+            Account::Github(v) => v.to_owned(),
+            Account::Web(v) => v.to_owned(),
+            Account::PGPFingerPrint(v) => String::from_utf8(v.to_vec()).unwrap(),
         }
     }
+
+    pub fn account_type(&self) -> &str {
+        match self {
+            Self::Discord(_) => "discord",
+            Self::Twitter(_) => "discord",
+            _ => "unkown",
+        }
+    }
+
+    /// constructs a [HashMap] of {Accout: VerifStatus} from a
+    /// [Vec<Account>] as `accounts` and a [VerifStatus] as `value`
+    pub fn into_hashmap(
+        accounts: Vec<Account>,
+        value: VerifStatus,
+    ) -> HashMap<Account, VerifStatus> {
+        let mut result = HashMap::new();
+        for account in accounts {
+            result.insert(account, value.clone());
+        }
+        result
+    }
 }
 
-/// TODO: move this to a "common" module
+// --------------------------------------
 #[derive(Debug, Serialize, Deserialize, Clone)]
-pub struct RegistrationRequest {
-    pub accounts: Vec<Account>,
-    pub id: AccountId32,
-    pub timeout: u64,
-    pub reg_index: u32,
+pub struct SubscribeAccountStateRequest {
+    pub version: String,
+    #[serde(rename = "type")]
+    pub _type: SubscribeAccountState,
+    pub payload: AccountId32,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct RequestedAccount {
+    /// wallet id
+    pub wallet_id: AccountId32,
+    pub field: AccountType,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct ChallengedAccount {
+    pub account: AccountId32,
+    pub field: AccountType,
+    pub challenge: String,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct VerificationRequest {
+    pub version: String,
+    #[serde(rename = "type")]
+    pub _type: SubscribeAccountState,
+    payload: RequestedAccount,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct VerificationResponse {
+    pub version: String,
+    #[serde(rename = "type")]
+    pub _type: RequestVerificationChallenge,
+    payload: Account,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub struct VerifyIdentityRequest {
+    pub version: String,
+    #[serde(rename = "type")]
+    pub _type: VerifyIdentity,
+    pub payload: ChallengedAccount,
+}
+
+// ------------------
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub enum SubscribeAccountState {
+    SubscribeAccountState,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub enum RequestVerificationChallenge {
+    RequestVerificationChallenge,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+pub enum VerifyIdentity {
+    VerifyIdentity,
+}
+
+#[derive(Debug, Serialize, Deserialize, Clone)]
+enum JsonResult {
+    JsonResult,
 }
 
+#[derive(Debug, Serialize, Deserialize, Clone)]
+enum NotifyAccountState {
+    NotifyAccountState,
```

Comment:
in dummyapi using in-memory, we could just keep track of subscribers in a HashMap and broadcast directly. with Redis, we need additional plumbing to:

1. Subscribe to Redis key changes 
2. Transform those Redis events into our WebSocket message format
3. Push these transformed events to relevant WebSocket clients

So rather than(dummyapi):
```
Client Subscribe -> Add to HashMap -> Direct broadcast
```

We need to split 2 steps:
```
Client Subscribe -> Add to HashMap 
Redis Change -> Transform to WebSocket format -> Look up relevant clients -> Push to each
```

so we'll need an additional service that:
1. Listens to Redis pub/sub 
2. Maps Redis keys to account IDs
3. Transforms state changes
4. Finds relevant WebSocket connections
5. Sends NotifyAccountState messages

```
use uuid::Uuid;
use futures_util::stream::SplitSink;
use tokio_tungstenite::WebSocketStream;

#[derive(Debug, Serialize)]
#[serde(rename_all = "camelCase")]
enum MessageType {
    SubscribeAccountState,
    NotifyAccountState,
    RequestVerificationSecret,
    VerifyIdentity,
    JsonResult,
}

#[derive(Debug, Serialize)]
struct WebSocketResponse {
    version: String,
    #[serde(rename = "type")]
    msg_type: MessageType,
    payload: WebSocketPayload,
}

#[derive(Debug, Serialize)]
#[serde(untagged)]
enum WebSocketPayload {
    JsonResult {
        #[serde(rename = "type")]
        result_type: String, // "ok" or "err"
        message: ResponseMessage,
    },
    NotifyState(NotifyAccountState),
}

#[derive(Debug, Serialize)]
#[serde(untagged)]
enum ResponseMessage {
    AccountState {
        account: AccountId32,
        hashed_info: String,
        verification_state: VerificationFields,
        pending_challenges: Vec<Vec<String>>,
    },
    Secret(String),
    VerificationResult(bool),
    Error(String),
}

#[derive(Debug, Clone)]
struct Listener {
    ip: [u8; 4],
    port: u16,
    redis_cfg: RedisConfig,
    clients: Arc<Mutex<HashMap<String, Arc<Mutex<SplitSink<WebSocketStream<TcpStream>, Message>>>>>>,
}

impl Listener {
    pub async fn new(websocket_cfg: WebsocketConfig, redis_cfg: RedisConfig) -> Self {
        Self {
            ip: websocket_cfg.ip,
            port: websocket_cfg.port,
            redis_cfg,
            clients: Arc::new(Mutex::new(HashMap::new()))
        }
    }

    async fn start_redis_subscription(&self, client_id: String, account: AccountId32) -> anyhow::Result<()> {
        let redis_client = RedisClient::open(self.redis_cfg.to_url()?)?;
        let mut pubsub = redis_client.get_async_pubsub().await?;
        let channel = format!("__keyspace@0__:{}:*", account);
        pubsub.subscribe(&channel).await?;

        while let Some(_) = pubsub.on_message().next().await {
            let mut conn = RedisConnection::create_conn(&self.redis_cfg)?;
            let state = conn.extract_info(&account)?;
            
            let response = WebSocketResponse {
                version: "1.0".to_string(),
                msg_type: MessageType::NotifyAccountState,
                payload: WebSocketPayload::NotifyState(NotifyAccountState {
                    account: account,
                    verification_state: state,
                }),
            };

            if let Some(client) = self.clients.lock().await.get(&client_id) {
                client.lock().await.send(Message::Text(serde_json::to_string(&response)?)).await?;
            }
        }
        Ok(())
    }

    pub async fn handle_incoming(&self, message: Message) -> anyhow::Result<WebSocketResponse> {
        match message {
            Message::Text(text) => {
                match serde_json::from_str::<SubscribeAccountStateRequest>(&text) {
                    Ok(request) => {
                        let mut conn = RedisConnection::create_conn(&self.redis_cfg)?;
                        if !conn.contains(&serde_json::to_string(&request.payload.to_owned())?) {
                            return Ok(WebSocketResponse {
                                version: "1.0".to_string(),
                                msg_type: MessageType::JsonResult,
                                payload: WebSocketPayload::JsonResult {
                                    result_type: "err".to_string(),
                                    message: ResponseMessage::Error(
                                        format!("account {} is not registered", request.payload)
                                    ),
                                },
                            });
                        }

                        let state = conn.extract_info(&request.payload)?;
                        let challenges = conn.get_challanges(&request.payload);
                        
                        Ok(WebSocketResponse {
                            version: "1.0".to_string(),
                            msg_type: MessageType::JsonResult,
                            payload: WebSocketPayload::JsonResult {
                                result_type: "ok".to_string(),
                                message: ResponseMessage::AccountState {
                                    account: request.payload,
                                    hashed_info: "TODO".to_string(),
                                    verification_state: state,
                                    pending_challenges: challenges,
                                },
                            },
                        })
                    }
                    Err(_) => {}
                }

                match serde_json::from_str::<VerificationRequest>(&text) {
                    Ok(request) => {
                        let mut conn = RedisConnection::create_conn(&self.redis_cfg)?;
                        let acc_type = request.payload.field;
                        
                        match conn.get_challenge_token_from_account_type(
                            &request.payload.wallet_id,
                            acc_type,
                        ) {
                            Some(token) => Ok(WebSocketResponse {
                                version: "1.0".to_string(),
                                msg_type: MessageType::JsonResult,
                                payload: WebSocketPayload::JsonResult {
                                    result_type: "ok".to_string(),
                                    message: ResponseMessage::Secret(token.show()),
                                },
                            }),
                            None => Ok(WebSocketResponse {
                                version: "1.0".to_string(),
                                msg_type: MessageType::JsonResult,
                                payload: WebSocketPayload::JsonResult {
                                    result_type: "err".to_string(),
                                    message: ResponseMessage::Error("could not find challenge".to_string()),
                                },
                            }),
                        }
                    }
                    Err(_) => {}
                }

                match serde_json::from_str::<VerifyIdentityRequest>(&text) {
                    Ok(request) => {
                        let mut conn = RedisConnection::create_conn(&self.redis_cfg)?;
                        match conn.get_challenge_token_from_account_type(
                            &request.payload.account,
                            request.payload.field,
                        ) {
                            Some(challenge) => {
                                let success = request.payload.challenge.eq(&challenge.show());
                                Ok(WebSocketResponse {
                                    version: "1.0".to_string(),
                                    msg_type: MessageType::JsonResult,
                                    payload: WebSocketPayload::JsonResult {
                                        result_type: "ok".to_string(),
                                        message: ResponseMessage::VerificationResult(success),
                                    },
                                })
                            }
                            None => Ok(WebSocketResponse {
                                version: "1.0".to_string(),
                                msg_type: MessageType::JsonResult,
                                payload: WebSocketPayload::JsonResult {
                                    result_type: "err".to_string(),
                                    message: ResponseMessage::Error("challenge not found".to_string()),
                                },
                            }),
                        }
                    }
                    Err(_) => {}
                }

                Ok(WebSocketResponse {
                    version: "1.0".to_string(),
                    msg_type: MessageType::JsonResult,
                    payload: WebSocketPayload::JsonResult {
                        result_type: "err".to_string(),
                        message: ResponseMessage::Error("unrecognized request format".to_string()),
                    },
                })
            }
            Message::Close(_) => Err(anyhow!("closing connection")),
            _ => Err(anyhow!("unrecognized message format"))
        }
    }

    pub async fn handle_connection(&self, stream: TcpStream) {
        let ws_stream = tokio_tungstenite::accept_async(stream)
            .await
            .expect("Error during websocket handshake");
        
        let client_id = Uuid::new_v4().to_string();
        let (outgoing, mut incoming) = ws_stream.split();
        let outgoing = Arc::new(Mutex::new(outgoing));
        
        self.clients.lock().await.insert(client_id.clone(), outgoing.clone());

        while let Some(Ok(message)) = incoming.next().await {
            match serde_json::from_str::<SubscribeAccountStateRequest>(&message.to_string()) {
                Ok(req) => {
                    let listener = self.clone();
                    let client_id = client_id.clone();
                    tokio::spawn(async move {
                        if let Err(e) = listener.start_redis_subscription(client_id, req.payload).await {
                            error!("Redis subscription error: {}", e);
                        }
                    });
                }
                Err(_) => {
                    match self.handle_incoming(message).await {
                        Ok(response) => {
                            if let Err(e) = outgoing.lock().await
                                .send(Message::Text(serde_json::to_string(&response)?))
                                .await 
                            {
                                error!("Failed to send response: {}", e);
                            }
                        }
                        Err(e) => error!("Failed to handle message: {}", e),
                    }
                }
            }
        }

        self.clients.lock().await.remove(&client_id);
    }

    pub async fn listen(self) -> anyhow::Result<()> {
        let addr = SocketAddr::from((self.ip, self.port));
        let listener = TcpListener::bind(&addr).await?;
        
        loop {
            match listener.accept().await {
                Ok((stream, addr)) => {
                    info!("incoming connection from {:?}...", addr);
                    let clone = self.clone();
                    tokio::spawn(async move {
                        clone.handle_connection(stream).await;
                    });
                }
                Err(e) => error!("could not accept ws connection! {}", e),
            }
        }
    }
}
```

Reply:
and split handle_incoming to process handlers
```
pub async fn handle_incoming(&self, message: Message) -> anyhow::Result<WebSocketResponse> {
        match message {
            Message::Text(text) => self.handle_text_message(text).await,
            Message::Close(_) => Err(anyhow!("closing connection")),
            _ => Err(anyhow!("unrecognized message format")),
        }
    }

    async fn handle_text_message(&self, text: String) -> anyhow::Result<WebSocketResponse> {
        if let Ok(request) = serde_json::from_str::<SubscribeAccountStateRequest>(&text) {
            return self.process_subscribe_account_state_request(request).await;
        }

        if let Ok(request) = serde_json::from_str::<VerificationRequest>(&text) {
            return self.process_verification_request(request).await;
        }

        if let Ok(request) = serde_json::from_str::<VerifyIdentityRequest>(&text) {
            return self.process_verify_identity_request(request).await;
        }

        Ok(self.error_response("unrecognized request format".to_string()))
    }
```